ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Outbound.hs

```diff
@@ -242,3 +244,19 @@ objectDiffusionOutbound tracer maxFifoLength ObjectPoolReader{..} _version =
     traceWith tracer (TraceObjectDiffusionOutboundSendMsgReplyObjects requestedObjects)
 
     pure (SendMsgReplyObjects requestedObjects (makeBundle st))
+
+-- | Retry an action that returns 'Maybe a' until it returns 'Just a'.
+retryMaybe :: Monad m => m (Maybe a) -> m a
+retryMaybe action = loop
+ where
+  loop = do
+    mx <- action
+    case mx of
+      Just x -> pure x
+      Nothing -> loop
+
+-- | Convert an empty map to 'Nothing', and a non-empty map to 'Just' that map.
+ensureNonEmpty :: Map k v -> Maybe (Map k v)
+ensureNonEmpty m
+  | Map.null m = Nothing
+  | otherwise = Just m
```

Comment:
Thought: from this point onward we could signal non-emptyness at the type level by passing around a [Data.NonEmpty.Map](https://hackage-content.haskell.org/package/nonempty-containers-0.3.5.0/docs/Data-Map-NonEmpty.html) from `nonempty-containers`.

Then. this function would become [Data.NonEmpty.Map.nonEmptyMap](https://hackage-content.haskell.org/package/nonempty-containers-0.3.5.0/docs/Data-Map-NonEmpty.html#v:nonEmptyMap)